app/scripts/controllers/permissions/selectors.test.js

```diff
@@ -0,0 +1,97 @@
+  describe('getChangedAccounts', () => {
+    it('returns the new value if the previous value is undefined', () => {
+      const newAccounts = new Map([['foo.bar', ['0x1']]]);
+      expect(getChangedAccounts(newAccounts)).toBe(newAccounts);
+    });
+
+    it('returns an empty map if the new and previous values are the same', () => {
+      const newAccounts = new Map([['foo.bar', ['0x1']]]);
+      expect(getChangedAccounts(newAccounts, newAccounts)).toStrictEqual(
+        new Map(),
+      );
+    });
+
+    it('returns a new map of the changed accounts if the new and previous values differ', () => {
+      // We set this on the new and previous value under the key 'foo.bar' to
+      // check that identical values are excluded.
+      const identicalValue = ['0x1'];
+
+      const previousAccounts = new Map([
+        ['bar.baz', ['0x1']], // included: different accounts
+        ['fizz.buzz', ['0x1']], // included: removed in new value
+      ]);
+      previousAccounts.set('foo.bar', identicalValue);
+
+      const newAccounts = new Map([
+        ['bar.baz', ['0x1', '0x2']], // included: different accounts
+        ['baz.fizz', ['0x3']], // included: brand new
+      ]);
+      newAccounts.set('foo.bar', identicalValue);
+
+      expect(getChangedAccounts(newAccounts, previousAccounts)).toStrictEqual(
+        new Map([
+          ['bar.baz', ['0x1', '0x2']],
+          ['fizz.buzz', []],
+          ['baz.fizz', ['0x3']],
+        ]),
+      );
+    });
+  });
+
+  describe('getPermittedAccountsByOrigin', () => {
+    it('memoizes and gets permitted accounts by origin', () => {
+      const state = {
+        subjects: {
+          'foo.bar': {
+            origin: 'foo.bar',
+            permissions: {
+              eth_accounts: {
+                caveats: [{ type: 'restrictReturnedAccounts', value: ['0x1'] }],
+              },
+            },
+          },
+          'bar.baz': {
+            origin: 'bar.baz',
+            permissions: {
+              eth_accounts: {
+                caveats: [{ type: 'restrictReturnedAccounts', value: ['0x2'] }],
+              },
+            },
+          },
+          'baz.bizz': {
+            origin: 'baz.fizz',
+            permissions: {
+              eth_accounts: {
+                caveats: [
+                  { type: 'restrictReturnedAccounts', value: ['0x1', '0x2'] },
+                ],
+              },
+            },
+          },
+          'no.accounts': {
+            // we shouldn't see this in the result
+            permissions: {
+              foobar: {},
+            },
+          },
+        },
+      };
+
+      const expected = new Map([
+        ['foo.bar', ['0x1']],
+        ['bar.baz', ['0x2']],
+        ['baz.fizz', ['0x1', '0x2']],
+      ]);
+
+      const received = getPermittedAccountsByOrigin(state);
+
+      expect(received).toStrictEqual(expected);
+      // The selector should return the memoized value if state.subjects is
```

Comment:
It might be worth checking that the memoization works by changing the state, and seeing that the answer changes.

Reply:
Agreed! 26d3949e2